pkg/gofr/datasource/file/ftp/file.go

```diff
@@ -0,0 +1,299 @@
+
+	return n, nil
+}
+
+// ReadAt reads data from the FTP file starting at the specified offset.
+func (f *file) ReadAt(p []byte, off int64) (n int, err error) {
+	defer f.postProcess(&FileLog{Operation: "ReadAt", Location: f.path}, time.Now())
+
+	resp, err := f.conn.RetrFrom(f.path, uint64(off))
+	if err != nil {
+		f.logger.Errorf("ReadAt failed. Error opening file with path %q at %v offset : %v", f.path, off, err)
+		return 0, err
+	}
+
+	defer resp.Close()
+
+	n, err = resp.Read(p)
+	if err != nil && !errors.Is(err, io.EOF) {
+		f.logger.Errorf("ReadAt failed. Error reading file with path %q at %v offset : %v", f.path, off, err)
+		return 0, err
+	}
+
+	f.logger.Logf("ReadAt successful. Read %v bytes from %q at offset of %v", n, f.path, off)
+
+	if n < len(p) {
+		return n, io.EOF
+	}
+
+	return n, nil
+}
+
+func (f *file) check(whence int, offset, length int64) (int64, error) {
+	switch whence {
+	case io.SeekStart:
+	case io.SeekEnd:
+		offset += length
+	case io.SeekCurrent:
+		offset += f.offset
+	default:
+		return 0, os.ErrInvalid
+	}
+
+	if offset < 0 || offset > length {
+		return 0, ErrOutOfRange
+	}
+
+	f.offset = offset
+
+	return f.offset, nil
+}
+
+// Seek sets the offset for the next Read or ReadAt operation.
+func (f *file) Seek(offset int64, whence int) (int64, error) {
+	defer f.postProcess(&FileLog{Operation: "Seek", Location: f.path}, time.Now())
+
+	n, err := f.conn.FileSize(f.path)
+	if err != nil {
+		f.logger.Errorf("Seek failed. Error : %v", err)
+		return 0, err
+	}
+
+	res, err := f.check(whence, offset, n)
+	if err != nil {
+		f.logger.Errorf("Seek failed. Error : %v", err)
+		return 0, err
+	}
+
+	f.logger.Logf("Seek successful. Offset at whence %v : %v", whence, res)
+
+	return res, nil
+}
+
+// Write writes data to the FTP file.
+func (f *file) Write(p []byte) (n int, err error) {
+	defer f.postProcess(&FileLog{Operation: "Write", Location: f.path}, time.Now())
+
+	reader := bytes.NewReader(p)
+
+	err = f.conn.StorFrom(f.path, reader, uint64(f.offset))
+	if err != nil {
+		f.logger.Errorf("Write failed. Error : %v", err)
+		return 0, err
+	}
+
+	f.offset += int64(len(p))
+
+	f.logger.Logf("Write successful. Wrote %v bytes to %v", len(p), f.path)
+
+	return len(p), nil
+}
+
+// WriteAt writes data to the FTP file starting at the specified offset.
+func (f *file) WriteAt(p []byte, off int64) (n int, err error) {
+	defer f.postProcess(&FileLog{Operation: "WriteAt", Location: f.path}, time.Now())
+
+	reader := bytes.NewReader(p)
+
+	err = f.conn.StorFrom(f.path, reader, uint64(off))
+	if err != nil {
+		f.logger.Errorf("WriteAt failed. Error writing in file with path %q at %v offset : %v", f.path, off, err)
+		return 0, err
+	}
+
+	f.logger.Logf("WriteAt success. Wrote %v bytes to %q at %v offset", len(p), f.path, off)
+
+	return len(p), nil
+}
+
+func (f *file) postProcess(fl *FileLog, startTime time.Time) {
+	duration := time.Since(startTime).Milliseconds()
+
+	fl.Duration = duration
+
+	f.logger.Debugf("%v", fl)
```

Comment:
This debug has no context such as text, it seems a left behind debug made for Dev purpose

Reply:
This is for logging the operation that has been carried out, in case we need it in a file etc. Although I guess Debugf is not required ...debug is suffice for this case.